### ISSUE TITLE
adding period range to the date_range

### ISSUE THREAD

**Issue by alonket** [pola-rs/polars#7525]
### Problem description

I'm dealing once in a while with dates and I need to create date ranges.
In pandas I can do it very easily:
`pd.date_range(pd.to_datetime(date) - pd.Timedelta(days=3), periods=3, freq="D").to_list()`

However when I tried to convert it to polars I had to deal with end_date.

Following my question in [SO](https://stackoverflow.com/questions/75722352/create-date-range-in-from-date-string-polars), the answer pointed out(and I agree) that it could be a good Idea to have something like this:
`date_range(start_date, period_of_time, interval)` in addition to `date_range(start_date, end_date, interval)`

**Comment by MarcoGorelli** [comment]
agree, being able to write `pl.date_range(date(2020, 1, 1), periods=3)` would be very convenient, I've found myself reaching for it a few times

**Comment by andreas-vester** [comment]
I totally agree. Introducing a ``periods`` argument is really helpful. See also my own question: https://stackoverflow.com/questions/78889556/create-date-range-with-predefined-number-of-periods-in-polars

This issue is open for quite some time now. I am wondering if there exist a concrete plan to incorporate it into the code base?

**Comment by MarcoGorelli** [comment]
@andreas-vester it's not high-priority, but if anyone from the community wanted to make a pull request, it would be welcome

**Comment by MarcoGorelli** [comment]
Made a start on this today:
```python
In [1]: import polars as pl

In [2]: from datetime import date

In [3]: pl.date_range(date(2020,1,1), periods=3, eager=True)
Out[3]: 
shape: (3,)
Series: 'literal' [date]
[
        2020-01-01
        2020-01-02
        2020-01-03
]

In [4]: pl.date_range(date(2020,1,1), date(2020, 1, 5), eager=True)
Out[4]: 
shape: (5,)
Series: 'literal' [date]
[
        2020-01-01
        2020-01-02
        2020-01-03
        2020-01-04
        2020-01-05
]
```
No promises as to when it'll be finished

**Comment by MarcoGorelli** [comment]
I've been struggling to make time for this - @mcrumiller would this interest you? It seems aligned with your interests

I'd given it a start in [this branch](https://github.com/MarcoGorelli/polars/tree/date-range-periods) though it's probably quite outdated by now

No worries / pressure of course, just pinging you in case it's of interest!

**Comment by mcrumiller** [comment]
I can take a look into it, sure. I'm currently working on something else but I agree this would definitely be useful.

**Comment by mcrumiller** [comment]
So this went a little stale, but I think it would still be useful. A few questions though, @MarcoGorelli would like your thoughts. Once `periods` is introduced, then of the four parameters start, end, interval, and periods, any three of them are necessary and sufficient to define a date/datetime range:

| start | end | interval | periods | description
| ------|-----|----------|---------|--------------
| X     | X   | X=i      | -       | evenly spaced periods of between start and end, each of length `i`
| X     | X   | -        | X=p     | `p` evenly spaced periods between start and end
| X     | -   | X=i      | X=n     | `n` evenly spaced periods starting at `start`, each of length `i`
| -     | X   | X        | X       | `n` evenly spaced periods ending at `end`, each of length `i`

Is this what we want to allow as well?

**Comment by MarcoGorelli** [comment]
thanks! yeah that seems right. and raise if all 4 are specified?

**Comment by mcrumiller** [comment]
Yeah, I think raise, even though it *is* possible to provide 4 valid args that work together.

**Comment by mcrumiller** [comment]
So, what do we do if we create a non-integer interval as a result of the number of periods? e.g. `pl.date_range(2025-01-01, 2025-01-10, periods=20)`? Should we treat this the same as numpy + ints, i.e. each date will simply be repeated twice?

**Comment by MarcoGorelli** [comment]
with `date_range` we currently get an error:
```python
In [26]: pl.date_range(date(2020,1,1), date(2020,1,10), '12h', eager=True)
---------------------------------------------------------------------------
ComputeError                              Traceback (most recent call last)
Cell In[26], line 1
----> 1 pl.date_range(date(2020,1,1), date(2020,1,10), '12h', eager=True)

File ~/scratch/.venv/lib/python3.12/site-packages/polars/functions/range/date_range.py:181, in date_range(start, end, interval, closed, eager)
    178 result = wrap_expr(plr.date_range(start_pyexpr, end_pyexpr, interval, closed))
    180 if eager:
--> 181     return F.select(result).to_series()
    183 return result

File ~/scratch/.venv/lib/python3.12/site-packages/polars/functions/lazy.py:2107, in select(eager, *exprs, **named_exprs)
   2051 """
   2052 Run polars expressions without a context.
   2053
   (...)   2104 └───────┘
   2105 """
   2106 empty_frame = pl.DataFrame() if eager else pl.LazyFrame()
-> 2107 return empty_frame.select(*exprs, **named_exprs)

File ~/scratch/.venv/lib/python3.12/site-packages/polars/dataframe/frame.py:9833, in DataFrame.select(self, *exprs, **named_exprs)
   9731 """
   9732 Select columns from this DataFrame.
   9733
   (...)   9826 └──────────────┘
   9827 """
   9828 from polars.lazyframe.opt_flags import QueryOptFlags
   9830 return (
   9831     self.lazy()
   9832     .select(*exprs, **named_exprs)
-> 9833     .collect(optimizations=QueryOptFlags._eager())
   9834 )

File ~/scratch/.venv/lib/python3.12/site-packages/polars/_utils/deprecation.py:97, in deprecate_streaming_parameter.<locals>.decorate.<locals>.wrapper(*args, **kwargs)
     93         kwargs["engine"] = "in-memory"
     95     del kwargs["streaming"]
---> 97 return function(*args, **kwargs)

File ~/scratch/.venv/lib/python3.12/site-packages/polars/lazyframe/opt_flags.py:330, in forward_old_opt_flags.<locals>.decorate.<locals>.wrapper(*args, **kwargs)
    327         optflags = cb(optflags, kwargs.pop(key))  # type: ignore[no-untyped-call,unused-ignore]
    329 kwargs["optimizations"] = optflags
--> 330 return function(*args, **kwargs)

File ~/scratch/.venv/lib/python3.12/site-packages/polars/lazyframe/frame.py:2332, in LazyFrame.collect(self, type_coercion, predicate_pushdown, projection_pushdown, simplify_expression, slice_pushdown, comm_subplan_elim, comm_subexpr_elim, cluster_with_columns, collapse_joins, no_optimization, engine, background, optimizations, **_kwargs)
   2330 # Only for testing purposes
   2331 callback = _kwargs.get("post_opt_callback", callback)
-> 2332 return wrap_df(ldf.collect(engine, callback))

ComputeError: `interval` input for `date_range` must consist of full days, got: 43200s
```

maybe something like that?

**Comment by mcrumiller** [comment]
Having (start, end, num_intervals) for `datetime_range` makes a lot of sense, but for `date_range` it feels a bit off but not *too* off. This is analogous to linspace/linear_space but for temporals with an integer return dtype, so I wonder if simply doing the numpy approach here works best (by returning days), as in:

```pycon
>>> import numpy as np
>>> np.linspace(1, 5, 6)
array([1. , 1.8, 2.6, 3.4, 4.2, 5. ])
>>> np.linspace(1, 5, 6, dtype="uint8")
array([1, 1, 2, 3, 4, 5], dtype=uint8)
```
Note the duplicate `1` value. Two alternatives are to 1) ensure that the span divided by `periods` is an integer, and raise otherwise (what you suggested), or 2) disallow (start, end, periods) input.

**Comment by MarcoGorelli** [comment]
> I wonder if simply doing the numpy approach here works best (by returning days)

sure, seems reasonable, thanks for looking into this!

**Comment by andreas-vester** [comment]
> disallow (start, end, periods) input.

I suggest to follow this route.

It's hard to come up with a use-case where you need all three input arguments. If you have ``start`` and ``end`` (and maybe ``interval``), the number of ``periods`` can be inferred. This is also true if you omit ``start`` or ``end`` and specify the number of ``periods``.

**Comment by mcrumiller** [comment]
> If you have start and end (and maybe interval), the number of periods can be inferred

The issue here is that the number of periods can be supplied, and the interval is then inferred, but the interval may not be a whole number of days. How to address this case is what we were discussing. This is analagous to calling `np.linspace` with a non-integer step but with an integer return dtype.

**Comment by etiennebacher** [comment]
From #23425 :

> This update adds the num_samples (I opted for this name over periods to be more in line with our range functions) parameter to date_range and datetime_range. 

I find this argument name very unclear about what it does. If you're looking for something that can be applied to all range functions, maybe `length_output` would do the job? `pl.date_range(start=date(1985, 1, 1), end=date(1985, 1, 10), length_output=5)` looks explicit enough to me.

(This was originally a comment in https://github.com/pola-rs/polars/pull/23425 but I deleted it there.)

**Comment by MarcoGorelli** [comment]
`num_samples` sounds fine to me, given it's already in `linear_space`
